### ISSUE TITLE
Not the best use of encapsulation

### ISSUE THREAD

**Issue by K4L1Ma** [kasperbe/port-processor#2]
Everything in the Domain is Public rendering constructors useless also ID is just a string Alias so we could easily instantiate a new port as 

```Go 
var port = Port{
 ID:ID("")
 Coordinates:Coordinates{{1},{2},{3}} << not a valid coordinate 
 {...}
}
```
ignoring all validations

* would like to understand a reason to have rawID and portInput as the NewPort Arguments why isn't rawID part of the portInputs?

**Comment by kasperbe** [comment]
> Everything in the Domain is Public rendering constructors useless also ID is just a string Alias so we could easily instantiate a new port as

I often find the public/private a debate in the Go community. Usually I have found relying on convention, and using `New` constructors to be good enough for most use cases, and ditching the getter/setter style. It tends to read nicer, however I do agree that sometimes the getter/setter approach is better, especially in libraries or code that is shared beyond just one team where you can't control conventions. 

I thought for this little exercise, convention would be fine. :) 


---

> would like to understand a reason to have rawID and portInput as the NewPort Arguments why isn't rawID part of the portInputs?

Simply kept these separate because the data was. No other reason. Could easily be merged into the `PortInput`, and probably should be. :)

**Comment by K4L1Ma** [comment]
Encapsulating and Getter/Setters are two different conversation, you can always Encapsulate and apply Tell don't ask principle to avoid Getter/Setters

**Comment by K4L1Ma** [comment]
actually as your in-memory impl. holds a copy of the port you don't even need to serialize anything or get anything out of the port itself other languages also ensures that domain entities comply with some serializable interface so we could expose its data but not interact with any domain logic, really useful when doing CQRS (Comand Query Responsibility Segregation) as you make sure your Read side never have access to your Command side

**Comment by kasperbe** [comment]
Maybe I don't fully understand, but doesn't Tell don't ask principle imply that I need to know how to perform what you're telling me to do? In this case, I do not want *any* architectural arrows pointing outwards from the domain, thus any serialization has to be done from the outside. 

One could of course argue that we can create simple DTOs and pass these around, but that would break domain safety, meaning you wouldn't know that this was a valid object, or that it had round tripped the domain layer.

For instance, in the repository:
```
func (r *Repository) Upsert(port domain.Port) {
	r.DB.Upsert(string(port.ID), port)
}
```

We could have an object that we could simply call `toStorage()` or something akin to that, however, since this is a domain object, knowing how to serialize for the storage engine, would break encapsulation as well, right?

**Comment by K4L1Ma** [comment]
you have different approaches on how to add serialization to an object in go, like complying to encode/Gob or as simple as complying to encode/Json by just implementing

```Go
func Marshal(v interface{}) ([]byte, error)]
func Unmarshal(data []byte, v interface{}) error]
```

with those interfaces, you could from within the domain without exposing the internal of the object map then to a map[string]interface{} you could store anywhere as you already have access to the data you could create your ORM DTOs and let the infra layer do its magic without it knowing anything about the Domain and Domain not knowing anything about the storage flavour

this is just a way of doing it, another would be using a [Design Pattern Called Memento](https://github.com/kamranahmedse/design-patterns-for-humans#-memento)

in Go, there is another more "dangerous" but valid approach where you create a DTOReplica of your DomainObject 
and map bidirectional A<=>B by just using the unsafe package but more on that in the future ;)

**Comment by kasperbe** [comment]
> with those interfaces, you could from within the domain without exposing the internal of the object map then to a map[string]interface{} you could store anywhere as you already have access to the data you could create your ORM DTOs and let the infra layer do its magic without it knowing anything about the Domain and Domain not knowing anything about the storage flavour

Sure. But that means you no longer know *anything* about the data. Do we need it in the example project? No.
But in real world scenarios, sometimes we want to lay out data differently on the storage than in the domain objects, thus we can't just map to an `interface{}`. This is what I tried to emulate, albeit, I may have been too implicit about it.

**Comment by K4L1Ma** [comment]
>Do we need it in the example project? No.

nah, to be honest, I am more keen on seeing [this](https://github.com/kasperbe/port-processor/issues/1)

>No.
But in real-world scenarios, sometimes we want to lay out data differently on the storage than in the domain objects, thus we can't just map to an interface{}

you won't be mapping to interface you would be either serializing as binary (good for things like kvsDBs) or to Map[string]interface{} which most ORMs already know how to handle for example MongoDB uses BSON which is an alias for Map[string]interface{} with extra functionality
